graph/sql/cockroach_test.go

```diff
@@ -3,34 +3,75 @@
 package sql
 
 import (
+	"fmt"
+	"math/rand"
+	"runtime"
+	"testing"
+	"time"
+	"unicode/utf8"
+
 	"github.com/cayleygraph/cayley/graph"
 	"github.com/cayleygraph/cayley/graph/graphtest"
 	"github.com/cayleygraph/cayley/graph/path/pathtest"
 	"github.com/cayleygraph/cayley/internal/dock"
 	"github.com/cayleygraph/cayley/quad"
+	"github.com/fsouza/go-dockerclient"
 	"github.com/lib/pq"
 	"github.com/stretchr/testify/require"
-	"testing"
-	"unicode/utf8"
 )
 
 func makeCockroach(t testing.TB) (graph.QuadStore, graph.Options, func()) {
 	var conf dock.Config // TODO
 
-	conf.Image = "cockroachdb/cockroach:beta-20170112"
+	conf.Image = "cockroachdb/cockroach:beta-20170413"
 	conf.Cmd = []string{"start", "--insecure"}
 
 	opts := graph.Options{"flavor": flavorCockroach}
 
-	addr, closer := dock.RunAndWait(t, conf, func(addr string) bool {
-		conn, err := pq.Open(`postgresql://root@` + addr + `:26257?sslmode=disable`)
-		if err != nil {
-			return false
+	var (
+		host   string
+		port   int
+		closer func()
+	)
+
+	if runtime.GOOS == "darwin" {
+		port = func() int {
+			rand.Seed(time.Now().Unix())
+			return rand.Intn(1000) + 45000
+		}()
+
+		conf.PortBindings = map[docker.Port][]docker.PortBinding{
+			"26257/tcp": []docker.PortBinding{
+				{
+					HostPort: fmt.Sprintf("%d", port),
+				},
+			},
 		}
-		conn.Close()
-		return true
-	})
-	db, err := connect(`postgresql://root@`+addr+`:26257?sslmode=disable`, flavorPostgres, nil)
+
+		host = "localhost"
+		_, closer = dock.RunAndWait(t, conf, func(string) bool {
```

Comment:
This is the same code block as on line 62. It needs to be a single code path with different variable values.

Also, you are doing this trick for darwin for SQL only. How about Mongo or any other non-trivial backend? This code really needs to be in `dock.RunAndWait` instead.